include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -166,34 +272,408 @@ struct __parallel_merge_submitter<_IdType, __internal::__optional_kernel_name<_N
+           |  rng[7]           |       |       |   +   |       |     |                        | slm[6]     |
+           |  rng[8]           |       |       |   +   |       |     |                        | slm[7]     |
+           |  rng[9]           |       |       |   +   |       |     |                        | slm[8]     |
+           |  rng[10]          |       |       |       |   +   |     |                        | slm[9]     |
+           |  rng[11]          |       |       |       |   +   |     |                        | slm[10]    |
+           |  rng[12]          |       |       |       |   +   |     |                        | slm[11]    |
+           |  .....            |       |       |       |       | +++ |                        | ...        |  
+           |  rng[M + 1]       |       |       |       |       |     |           +            | slm[M]     |  
+           |  rng[M + 2]       |       |       |       |       |     |           +            | slm[M + 1] |  
+           |  rng[M + 3]       |       |       |       |       |     |           +            | slm[M + 2] |  
+           |  rng[M + 4]       |       |       |       |       |     |                        | slm[M + 3] |
+           |  rng[M + 5]       |       |       |       |   -   |     |                        |            |  <--- __idx_global_end
+           |  .....            |       |       |       |       | --- |                        |            |  
+           |  rng[M + M + 1]   |       |       |       |       |     |           -            |            |  
+           +-------------------+--------------------------------------------------------------+------------+
+                                                   ^
+                                                   |
+                                              __local_id
+           
+            "+" - load one source data item ito SLM
+        */
+
+        const std::size_t __wg_data_size_rng = __idx_global_end - __idx_global_begin;
+        if (__wg_data_size_rng > 0)
+        {
+            // Calculate the size of the current part of merging data per work-item
+            const std::size_t __loading_data_per_wi = oneapi::dpl::__internal::__dpl_ceiling_div(__rng.size(), __wi_in_one_wg);
+            assert(__loading_data_per_wi > 0);
+
+            if (__loading_data_per_wi > 1)
+            {
+                const auto __slm_idx_begin = __local_id * __loading_data_per_wi;
+                const auto __slm_idx_end = __slm_idx_begin + __loading_data_per_wi;
+
+                std::size_t __slm_idx = __slm_idx_begin;
+                std::size_t __rng_idx = __idx_global_begin + __slm_idx;
+
+                _ONEDPL_PRAGMA_UNROLL
+                for (; __slm_idx < __slm_idx_end && __rng_idx < __idx_global_end; ++__slm_idx, ++__rng_idx)
+                    __slm[__slm_idx] = __rng[__rng_idx];
```

Comment:
I believe the global memory access pattern here from `__rng` is poor. This may be a big bottleneck. For a good access pattern, we want all items in a sub-group to access adjacent indices from `__rng` in each iteration of the loop.

My thoughts are that we should ensure good global access patterns first, and then profile and optimize SLM accesses to minimize bank conflicts.